### PR TITLE
allow to generate cert/csr with ECDSA key

### PR DESCRIPTION
**Pull request by vktg**
- [ ] Redmine Issue: https://redmine.pfsense.org/issues/9843
- [ ] Ready for review

Adds ability to generate certificates/CSRs with ECDSA keys.

### CODE REVIEW

**Comment by vktg** [comment]
$openssl_eckeys from 'openssl ecparam -list_curves' of pfSense 2.5 2.5.0.a.20191022.1108 ( OpenSSL 1.1.1a-freebsd  20 Nov 2018 )

File: src/etc/inc/certs.inc [hunk 1]
```diff
@@ -27,7 +27,9 @@ define("OPEN_SSL_CONF_PATH", "/etc/ssl/openssl.cnf");
...
...ecp160r2", "secp192k1", "secp224k1", "secp224r1", "secp256k1", "secp384r1", "secp521r1", "prime192v1", "prime192v2", "prime192v3", "prime239v1", "prime239v2", "prime239v3", "prime256v1", "sect113r1", "sect113r2", "sect131r1", "sect131r2", "sect163k1", "sect163r1", "sect163r2", "sect193r1", "sect193r2", "sect233k1", "sect233r1", "sect239k1", "sect283k1", "sect283r1", "sect409k1", "sect409r1", "sect571k1", "sect571r1", "c2pnb163v1", "c2pnb163v2", "c2pnb163v3", "c2pnb176v1", "c2tnb191v1", "c2tnb191v2", "c2tnb191v3", "c2pnb208w1", "c2tnb239v1", "c2tnb239v2", "c2tnb239v3", "c2pnb272w1", "c2pnb304w1", "c2tnb359v1", "c2pnb368w1", "c2tnb431r1", "wap-wsg-idm-ecid-wtls1", "wap-wsg-idm-ecid-wtls3", "wap-wsg-idm-ecid-wtls4", "wap-wsg-idm-ecid-wtls5", "wap-wsg-idm-ecid-wtls6", "wap-wsg-idm-ecid-wtls7", "wap-wsg-idm-ecid-wtls8", "wap-wsg-idm-ecid-wtls9", "wap-wsg-idm-ecid-wtls10", "wap-wsg-idm-ecid-wtls11", "wap-wsg-idm-ecid-wtls12", "brainpoolP160r1", "brainpoolP160t1", "brainpoolP192r1", "brainpoolP192t1", "brainpoolP224r1", "brainpoolP224t1", "brainpoolP256r1", "brainpoolP256t1", "brainpoolP320r1", "brainpoolP320t1", "brainpoolP384r1", "brainpoolP384t1", "brainpoolP512r1", "brainpoolP512t1");
```
Comment (jim-p) [inline src/etc/inc/certs.inc]: Rather than hardcode this list, it should use `openssl_get_curve_names()` to dynamically populate the content with whatever the system supports at the time.

Additionally, these are not "keys" but curves. EC is a type of key, but the value selected by the user is a specific elliptic curve name, so the variable should be `$openssl_ecnames`, and all related references need to be changed to reflect that (e.g. `$eckey` should be `$ecname`, labels fixed, and so on).

File: src/etc/inc/certs.inc [hunk 2]
```diff
@@ -312,7 +314,7 @@ function cert_import(& $cert, $crt_str, $key_str) {
 	return true;
 }
 
-function cert_create(& $cert, $caref, $keylen, $lifetime, $dn, $type = "user", $digest_alg = "sha256") {
+function cert_create(& $cert, $caref, $keylen, $lifetime, $dn, $type = "user", $digest_alg = "sha256", $eckey = "brainpoolP160r1", $keytype = "RSA") {
```
Comment (jim-p) [inline src/etc/inc/certs.inc]: I'd prefer to see `$keytype` first, and then `$ecname` (not `$eckey`, see above comment) after, but that's not a hard requirement.

File: src/etc/inc/certs.inc [hunk 3]
```diff
@@ -408,11 +415,15 @@ function csr_generate(& $cert, $keylen, $dn, $type = "user", $digest_alg = "sha2
 
 	$args = array(
 		"x509_extensions" => $cert_type,
-		"req_extensions" => "req_{$cert_type}",
```
Comment (jim-p) [inline src/etc/inc/certs.inc]: Why is this line removed? It doesn't appear to be related to the other changes.

File: src/usr/local/www/system_certmanager.php [hunk 4]
```diff
@@ -772,11 +794,31 @@ function list_csrs() {
...
+	$group->add(new Form_Select(
 		'keylen',
-		'*Key length',
+		null,
 		$pconfig['keylen'],
 		array_combine($cert_keylens, $cert_keylens)
 	));
+	$section->add($group);
+
+	$group = new Form_Group($i == 0 ? '*EC key':'');
```
Comment (jim-p) [inline src/usr/local/www/system_certmanager.php]: "EC Key" is very user unfriendly. It should be spelled out ("Elliptic Curve Name")

File: src/usr/local/www/system_certmanager.php [hunk 5]
```diff
@@ -850,11 +892,31 @@ function list_csrs() {
...
+	$group->add(new Form_Select(
 		'csr_keylen',
-		'*Key length',
+		null,
 		$pconfig['csr_keylen'],
 		array_combine($cert_keylens, $cert_keylens)
 	));
+	$section->add($group);
+
+	$group = new Form_Group($i == 0 ? '*EC key':'');
```
Comment (jim-p) [inline src/usr/local/www/system_certmanager.php]: See previous note about the label.

File: src/usr/local/www/system_certmanager.php [hunk 6]
```diff
@@ -850,11 +892,31 @@ function list_csrs() {
...
 
 	$section->addInput(new Form_Select(
+		'csr_keytype',
+		'*Key type',
+		$pconfig['csr_keytype'],
+		array_combine($cert_keytypes, $cert_keytypes)
+	));
+
+	$group = new Form_Group($i == 0 ? '*Key length':'');
+        $group->addClass('csr_rsakeys');
```
Comment (jim-p) [inline src/usr/local/www/system_certmanager.php]: Something is off with the whitespace on this line

File: src/usr/local/www/system_certmanager.php [hunk 7]
```diff
@@ -772,11 +794,31 @@ function list_csrs() {
...
 
 	$section->addInput(new Form_Select(
+		'keytype',
+		'*Key type',
+		$pconfig['keytype'],
+		array_combine($cert_keytypes, $cert_keytypes)
+	));
+
+	$group = new Form_Group($i == 0 ? '*Key length':'');
+        $group->addClass('rsakeys');
```
Comment (jim-p) [inline src/usr/local/www/system_certmanager.php]: Something is off with the whitespace on this line

File: src/usr/local/www/system_certmanager.php [hunk 8]
```diff
@@ -335,17 +345,29 @@
...
 					break;
 				case "external":
+					if (isset($_POST["csr_keytype"]) && !in_array($_POST["csr_keytype"], $cert_keytypes)) {
+						array_push($input_errors, gettext("Please select a valid Key Type."));
+					}
 					if (isset($_POST["csr_keylen"]) && !in_array($_POST["csr_keylen"], $cert_keylens)) {
 						array_push($input_errors, gettext("Please select a valid Key Length."));
 					}
+					if (isset($_POST["csr_eckey"]) && !in_array($_POST["csr_eckey"], $openssl_eckeys)) {
+						array_push($input_errors, gettext("Please select a valid EC Key."));
```
Comment (jim-p) [inline src/usr/local/www/system_certmanager.php]: Make sure this is changed to match the label.

File: src/usr/local/www/system_certmanager.php [hunk 9]
```diff
@@ -335,17 +345,29 @@
...
 			switch ($pconfig['method']) {
 				case "internal":
+					if (isset($_POST["keytype"]) && !in_array($_POST["keytype"], $cert_keytypes)) {
+						array_push($input_errors, gettext("Please select a valid Key Type."));
+					}
 					if (isset($_POST["keylen"]) && !in_array($_POST["keylen"], $cert_keylens)) {
 						array_push($input_errors, gettext("Please select a valid Key Length."));
 					}
+					if (isset($_POST["eckey"]) && !in_array($_POST["eckey"], $openssl_eckeys)) {
+						array_push($input_errors, gettext("Please select a valid EC Key."));
```
Comment (jim-p) [inline src/usr/local/www/system_certmanager.php]: Make sure this is changed to match the label.

File: src/etc/inc/certs.inc (hunk 2 above)
Comment (jim-p) [inline src/etc/inc/certs.inc]: Why is `brainpoolP160r1` the default? We should probably at least use `brainpoolP256r1` for a default, since [NIST recommends using a minimum of 256 bit curves](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-57pt1r4.pdf).

File: src/etc/inc/certs.inc [hunk 10]
```diff
@@ -394,7 +401,7 @@ function cert_create(& $cert, $caref, $keylen, $lifetime, $dn, $type = "user", $
 	return true;
 }
 
-function csr_generate(& $cert, $keylen, $dn, $type = "user", $digest_alg = "sha256") {
+function csr_generate(& $cert, $keylen, $dn, $type = "user", $digest_alg = "sha256", $eckey = "brainpoolP160r1", $keytype = "RSA") {
```
Comment (jim-p) [inline src/etc/inc/certs.inc]: See previous comment about the default curve.

File: src/usr/local/www/system_certmanager.php [hunk 11]
```diff
@@ -93,9 +95,13 @@
 
 if ($act == "new") {
 	$pconfig['method'] = $_POST['method'];
+	$pconfig['keytype'] = "RSA";
 	$pconfig['keylen'] = "2048";
+	$pconfig['eckey'] = "brainpoolP160r1";
```
Comment (jim-p) [inline src/usr/local/www/system_certmanager.php]: See previous comment about the default curve.

File: src/usr/local/www/system_certmanager.php [hunk 12]
```diff
@@ -93,9 +95,13 @@
...
 	$pconfig['digest_alg'] = "sha256";
+	$pconfig['csr_keytype'] = "RSA";
 	$pconfig['csr_keylen'] = "2048";
+	$pconfig['csr_eckey'] = "brainpoolP160r1";
```
Comment (jim-p) [inline src/usr/local/www/system_certmanager.php]: See previous comment about the default curve.

File: src/usr/local/www/system_certmanager.php (hunk 11 above)
Comment (vktg) [reply src/usr/local/www/system_certmanager.php]: fixed

File: src/etc/inc/certs.inc (hunk 10 above)
Comment (vktg) [reply src/etc/inc/certs.inc]: fixed

File: src/usr/local/www/system_certmanager.php (hunk 9 above)
Comment (vktg) [reply src/usr/local/www/system_certmanager.php]: fixed

File: src/usr/local/www/system_certmanager.php (hunk 4 above)
Comment (vktg) [reply src/usr/local/www/system_certmanager.php]: changed to Elliptic Curve Name

File: src/etc/inc/certs.inc (hunk 3 above)
Comment (vktg) [reply src/etc/inc/certs.inc]: fixed
typo

File: src/etc/inc/certs.inc (hunk 2 above)
Comment (vktg) [reply src/etc/inc/certs.inc]: fixed

File: src/etc/inc/certs.inc (hunk 1 above)
Comment (vktg) [reply src/etc/inc/certs.inc]: changed to openssl_get_curve_names() and all $eckey renamed to $ecname

**Comment by vktg** [comment]
After completing all issues, I want to make the same changes to system_camanager.php (so as not to create an additional PR)

File: src/usr/local/www/system_certmanager.php [hunk 13]
```diff
@@ -772,11 +794,31 @@ function list_csrs() {
...
 		'keylen',
-		'*Key length',
+		null,
 		$pconfig['keylen'],
 		array_combine($cert_keylens, $cert_keylens)
 	));
+	$section->add($group);
+
+	$group = new Form_Group($i == 0 ? '*Elliptic Curve Name':'');
+        $group->addClass('ecnames');
```
Comment (jim-p) [inline src/usr/local/www/system_certmanager.php]: Change leading spaces to tab

File: src/usr/local/www/system_certmanager.php [hunk 14]
```diff
@@ -850,11 +892,31 @@ function list_csrs() {
...
 		'csr_keylen',
-		'*Key length',
+		null,
 		$pconfig['csr_keylen'],
 		array_combine($cert_keylens, $cert_keylens)
 	));
+	$section->add($group);
+
+	$group = new Form_Group($i == 0 ? '*Elliptic Curve Name':'');
+        $group->addClass('csr_ecnames');
```
Comment (jim-p) [inline src/usr/local/www/system_certmanager.php]: Change leading spaces to tab

File: src/usr/local/www/system_certmanager.php (hunk 13 above)
Comment (vktg) [reply src/usr/local/www/system_certmanager.php]: fixed

File: src/usr/local/www/system_certmanager.php (hunk 14 above)
Comment (vktg) [reply src/usr/local/www/system_certmanager.php]: fixed

**Comment by vktg** [comment]
same changes for system_camanager.php

**Comment by vktg** [comment]
done with info blocks

**Comment by vktg** [comment]
do we need such info blocks on system_camanager.php page?

**Comment by jim-p** [comment]
> do we need such info blocks on system_camanager.php page?

Yes,  but not on this PR. I'm working on something else for that page and I'll take care of it.

**Comment by jim-p** [comment]
Looks good now from some limited testing, enough to merge it and test it more thoroughly.
